File: node/tests/mime.test.ts
import { describe, it, expect } from 'vitest';
import { buildRfc5322Email } from '../src/mime.js';
import type { MimeSendEmailOptions } from '../src/mime.js';

describe('buildRfc5322Email', () => {
  it('produces valid RFC 5322 for simple text email', () => {
    const opts: MimeSendEmailOptions = {
      to: 'recipient@hai.ai',
      subject: 'Test Subject',
      body: 'Hello, world!',
    };
    const raw = buildRfc5322Email(opts, 'sender@hai.ai');
    const text = raw.toString('utf-8');

    expect(text).toContain('From: <sender@hai.ai>\r\n');
    expect(text).toContain('To: recipient@hai.ai\r\n');
    expect(text).toContain('Subject: Test Subject\r\n');
    expect(text).toContain('Date: ');
    expect(text).toContain('Message-ID: <');
    expect(text).toContain('MIME-Version: 1.0\r\n');
    expect(text).toContain('Content-Type: text/plain; charset=utf-8\r\n');
    expect(text).toContain('Hello, world!');
  });

  it('handles attachments', () => {
    const opts: MimeSendEmailOptions = {
      to: 'recipient@hai.ai',
      subject: 'With Attachments',
      body: 'See attached.',
      attachments: [
        {
          filename: 'file1.txt',
          contentType: 'text/plain',
          data: Buffer.from('content of file 1'),
        },
        {
          filename: 'file2.pdf',
          contentType: 'application/pdf',
          data: Buffer.from('fake pdf content'),
        },
      ],
    };

    const raw = buildRfc5322Email(opts, 'sender@hai.ai');
    const text = raw.toString('utf-8');

    expect(text).toContain('Content-Type: multipart/mixed; boundary=');
    expect(text).toContain('Content-Disposition: attachment; filename="file1.txt"');
    expect(text).toContain('Content-Disposition: attachment; filename="file2.pdf"');
    expect(text).toContain('Content-Transfer-Encoding: base64');
    expect(text).toContain('See attached.');
  });

  it('handles reply threading', () => {
    const opts: MimeSendEmailOptions = {
      to: 'recipient@hai.ai',
      subject: 'Re: Original',
      body: 'Reply body',
      inReplyTo: '<original-id@hai.ai>',
    };

    const raw = buildRfc5322Email(opts, 'sender@hai.ai');
    const text = raw.toString('utf-8');

    expect(text).toContain('In-Reply-To: <original-id@hai.ai>\r\n');
    expect(text).toContain('References: <original-id@hai.ai>\r\n');
  });

  it('sanitizes CRLF injection', () => {
    const opts: MimeSendEmailOptions = {
      to: 'recipient@hai.ai',
      subject: 'Bad\r\nBcc: attacker@evil.com',
      body: 'Body',
    };

    const raw = buildRfc5322Email(opts, 'sender@hai.ai');
    const text = raw.toString('utf-8');

    // No line should start with "Bcc:" (CRLF injection prevented)
    const lines = text.split('\r\n');
    for (const line of lines) {
      expect(line.startsWith('Bcc:')).toBe(false);
    }
    // Subject should be sanitized
    expect(text).toContain('Subject: BadBcc: attacker@evil.com\r\n');
  });

  it('uses CRLF line endings', () => {
    const opts: MimeSendEmailOptions = {
      to: 'recipient@hai.ai',
      subject: 'Test',
      body: 'Body',
    };

    const raw = buildRfc5322Email(opts, 'sender@hai.ai');
    const text = raw.toString('utf-8');

    expect(text).toContain('\r\n');
    // Split by CRLF, no remaining bare \n in any part (except within body text)
    const headerSection = text.split('\r\n\r\n')[0];
    const headerParts = headerSection.split('\r\n');
    for (const part of headerParts) {
      expect(part).not.toContain('\n');
    }
  });
});

File: go/mime_test.go
package haiai

import (
	"bytes"
	"net/mail"
	"strings"
	"testing"
)

func TestBuildSimpleTextEmail(t *testing.T) {
	opts := SendEmailOptions{
		To:      "recipient@hai.ai",
		Subject: "Test Subject",
		Body:    "Hello, world!",
	}

	raw, err := BuildRFC5322Email(opts, "sender@hai.ai")
	if err != nil {
		t.Fatalf("BuildRFC5322Email failed: %v", err)
	}

	msg, err := mail.ReadMessage(bytes.NewReader(raw))
	if err != nil {
		t.Fatalf("failed to parse email: %v", err)
	}

	if got := msg.Header.Get("To"); got != "recipient@hai.ai" {
		t.Errorf("To = %q, want %q", got, "recipient@hai.ai")
	}
	if got := msg.Header.Get("Subject"); got != "Test Subject" {
		t.Errorf("Subject = %q, want %q", got, "Test Subject")
	}
	if got := msg.Header.Get("Date"); got == "" {
		t.Error("Date header is empty")
	}
	if got := msg.Header.Get("Message-Id"); got == "" {
		t.Error("Message-ID header is empty")
	}
	if got := msg.Header.Get("Mime-Version"); got != "1.0" {
		t.Errorf("MIME-Version = %q, want %q", got, "1.0")
	}
}

func TestBuildEmailWithAttachments(t *testing.T) {
	opts := SendEmailOptions{
		To:      "recipient@hai.ai",
		Subject: "With Attachments",
		Body:    "See attached.",
		Attachments: []EmailAttachment{
			{
				Filename:    "file1.txt",
				ContentType: "text/plain",
				Data:        []byte("content of file 1"),
			},
			{
				Filename:    "file2.pdf",
				ContentType: "application/pdf",
				Data:        []byte("fake pdf content"),
			},
		},
	}

	raw, err := BuildRFC5322Email(opts, "sender@hai.ai")
	if err != nil {
		t.Fatalf("BuildRFC5322Email failed: %v", err)
	}

	text := string(raw)
	if !strings.Contains(text, "Content-Type: multipart/mixed; boundary=") {
		t.Error("missing multipart/mixed content type")
	}
	if !strings.Contains(text, `Content-Disposition: attachment; filename="file1.txt"`) {
		t.Error("missing file1.txt attachment")
	}
	if !strings.Contains(text, `Content-Disposition: attachment; filename="file2.pdf"`) {
		t.Error("missing file2.pdf attachment")
	}
	if !strings.Contains(text, "Content-Transfer-Encoding: base64") {
		t.Error("missing base64 transfer encoding")
	}
	if !strings.Contains(text, "See attached.") {
		t.Error("missing body text")
	}
}

func TestBuildReplyEmail(t *testing.T) {
	opts := SendEmailOptions{
		To:        "recipient@hai.ai",
		Subject:   "Re: Original",
		Body:      "Reply body",
		InReplyTo: "<original-id@hai.ai>",
	}

	raw, err := BuildRFC5322Email(opts, "sender@hai.ai")
	if err != nil {
		t.Fatalf("BuildRFC5322Email failed: %v", err)
	}

	msg, err := mail.ReadMessage(bytes.NewReader(raw))
	if err != nil {
		t.Fatalf("failed to parse email: %v", err)
	}

	if got := msg.Header.Get("In-Reply-To"); got != "<original-id@hai.ai>" {
		t.Errorf("In-Reply-To = %q, want %q", got, "<original-id@hai.ai>")
	}
	if got := msg.Header.Get("References"); got != "<original-id@hai.ai>" {
		t.Errorf("References = %q, want %q", got, "<original-id@hai.ai>")
	}
}

func TestCRLFInjectionSanitized(t *testing.T) {
	opts := SendEmailOptions{
		To:      "recipient@hai.ai",
		Subject: "Bad\r\nBcc: attacker@evil.com",
		Body:    "Body",
	}

	raw, err := BuildRFC5322Email(opts, "sender@hai.ai")
	if err != nil {
		t.Fatalf("BuildRFC5322Email failed: %v", err)
	}

	text := string(raw)
	// No line should start with "Bcc:" (CRLF injection prevented)
	for _, line := range strings.Split(text, "\r\n") {
		if strings.HasPrefix(line, "Bcc:") {
			t.Error("CRLF injection succeeded: found header line starting with Bcc:")
		}
	}
	// Subject should be sanitized
	if !strings.Contains(text, "Subject: BadBcc: attacker@evil.com\r\n") {
		t.Error("subject was not properly sanitized")
	}
}

func TestCRLFLineEndings(t *testing.T) {
	opts := SendEmailOptions{
		To:      "recipient@hai.ai",
		Subject: "Test",
		Body:    "Body",
	}

	raw, err := BuildRFC5322Email(opts, "sender@hai.ai")
	if err != nil {
		t.Fatalf("BuildRFC5322Email failed: %v", err)
	}

	text := string(raw)
	if !strings.Contains(text, "\r\n") {
		t.Error("output does not contain CRLF line endings")
	}
}
